browser/static/js/canvas.js

```diff
@@ -215,3 +171,86 @@ class CanvasState {
     return this.isPressed;
   }
 }
+
+class Pan {
```

Comment:
Did you consider a shared parent `Action` class?

I don't think JavaScript actually has a proper `Interface`, unfortunately, but TypeScript does (for the future).

Reply:
I've given this an attempt with commit 5e830e5e2e8bc2bb63575f7e8d8f9206943c4d72, let me know if you think this looks better. There's not much logic in the Action class, it just has docstrings and a warning that the function is not implemented.

Reply:
That looks great! You may want to consider making it an abstract class, like in [this SO post](https://stackoverflow.com/a/30560792). It looks like they try to check if a particular method is defined, and throws an error otherwise, which could be helpful here for the 3 required methods. (which is sort of a hacky implementation of Java's abstract classes and interfaces).